### PR TITLE
Fix depleting bytes from the response.

### LINKED ISSUE #225: Chucker corrupts image resources.
## :loudspeaker: Describe the bug
If `OkHttp` client with the `ChuckerInterceptor` is used for image downloads then the requests fail.

## :bomb: To Reproduce
1. Run the sample project. - https://github.com/MiSikora/Chucker-Bug
2. Run the app.
3. Select the check box to enable Chucker.
4. Press either `Coil` or `Picasso` buttons.

### :wrench: Expected behavior
The image should be downloaded and displayed in the ImageView. It should work in the same way as if the `ChuckerInterceptor` was disabled.

## :camera: Screenshots
Stack trace for `Picasso`.

```
java.io.IOException: Failed to decode stream.
  at com.squareup.picasso.BitmapHunter.decodeStream(BitmapHunter.java:161)
  at com.squareup.picasso.BitmapHunter.hunt(BitmapHunter.java:229)
  at com.squareup.picasso.RequestCreator.get(RequestCreator.java:429)
  at io.mehow.demo.MainActivity$onCreate$3$1.invoke(MainActivity.kt:54)
  at io.mehow.demo.MainActivity$onCreate$3$1.invoke(MainActivity.kt:23)
  at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:30)
```

Stack trace for `Coil`.

```
java.lang.IllegalStateException: BitmapFactory returned a null Bitmap. Often this means BitmapFactory could not decode the image data read from the input source (e.g. network or disk) as it's not encoded as a valid image format.
  at coil.decode.BitmapFactoryDecoder.decode(BitmapFactoryDecoder.kt:156)
  at coil.RealImageLoader$loadData$2.invokeSuspend(RealImageLoader.kt:392)
  at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
  at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
  at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:561)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:727)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:667)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:655)
```

## :iphone: Smartphone
 - Device: Any Device
 - OS: Any OS
 - Lib version: `3.1.1`

## :page_facing_up: Additional context
Issue does not exist in `3.1.0`. It was introduced with `3.1.1`.

### PR DESCRIPTION
**Pull request by MiSikora**
## :page_facing_up: Context
*Why did you change something? Is there an [issue](https://github.com/ChuckerTeam/chucker/issues) to link here? Or an extarnal link?*

There is an issue with Chucker and images. https://github.com/ChuckerTeam/chucker/issues/225

## :pencil: Changes
*Which code did you change? How?*

Major focus was on `processResponseBody` method of `ChuckerInterceptor`.
* If source is identified as an image do not read bytes directly but copy them instead.
* Simplify gzip logic.

Minor change was to use `HttpUrl` directly instead of relying on `android.net.Uri`. It was unnecessary and made JVM tests harder. Aditional benefit is that now https://github.com/ChuckerTeam/chucker/issues/155 is simpler to implement.

## :warning: Breaking
*Is there something breaking in the API? Any class or method signature changed?*

No API breaking changes. Resulting Response does not use `FixedLengthSource` anymore as bytes are copied to the internal buffer but it shouldn't matter to the end consumers.

## :pencil: How to test
*Is there a special case to test your changes?*

I added jUnit tests but for manuall testing you can run https://github.com/MiSikora/Chucker-Bug with fixes from this PR.

Also, sorry if something is messed up in tests. I'm not used to testing with mocks but I didn't see any way around it since interceptor is coupled with the Android framework.

## :crystal_ball: Next steps
*Do we have to plan something else after the merge?*

Hotfix release would be welcome. :)

Closes #225

### CODE REVIEW

File: library/src/main/java/com/chuckerteam/chucker/api/ChuckerInterceptor.kt [hunk 1]
```diff
@@ -178,6 +182,9 @@ class ChuckerInterceptor @JvmOverloads constructor(
         return builder.build()
     }
 
+    /** Checks if the OkHttp response uses gzip encoding. */
+    private val Response.isGzipped get() = io.bodyIsGzipped(headers()[CONTENT_ENCODING])
```
Comment (vbuberen) [inline library/src/main/java/com/chuckerteam/chucker/api/ChuckerInterceptor.kt]: I suggest to move it to OkHttpUtils to other similar extensions.

**Comment by vbuberen** [comment]
Thanks for your contribution and for adding tests.
Just a minor comment and we are good to go.

I will release a hotfix release `3.1.2` after merging your PR.

**Comment by MiSikora** [comment]
During refactoring i removed `Response.hasBody()` extension and rely durint interception solely on `OkHttp` with check on `body()` whether is null or not. I'm not sure if the mentioned extension served any purpose as a workaround etc. Please let know if this should be reverted.

File: library/src/main/java/com/chuckerteam/chucker/internal/support/OkHttpUtils.kt [hunk 2]
```diff
@@ -2,39 +2,38 @@
...
-    return this.contentLenght != -1L || this.isChunked
-}
+/** Checks if the OkHttp response uses gzip encoding. */
+internal val Response.isGzipped: Boolean
+    get() {
+        return this.headers().containsGzip
+    }
 
-internal val Response.contentLenght: Long
+/** Checks if the OkHttp response uses gzip encoding. */
```
Comment (vbuberen) [inline library/src/main/java/com/chuckerteam/chucker/internal/support/OkHttpUtils.kt]:
```suggestion
/** Checks if the OkHttp request uses gzip encoding. */
```

**Comment by vbuberen** [comment]
> While refactoring I removed `Response.hasBody()` extension and rely during interception solely on `OkHttp` with a null check on `body()`. I'm not sure if the mentioned extension served any purpose as a workaround etc. Please let me know if this should be reverted.

That's fine from my point of view.

**Comment by vbuberen** [comment]
LGTM 🚢

**Pull request merged**